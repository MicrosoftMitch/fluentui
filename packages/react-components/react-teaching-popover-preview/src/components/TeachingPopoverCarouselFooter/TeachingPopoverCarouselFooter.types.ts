import * as React from 'react';
import { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import { Button } from '@fluentui/react-button';
import { PopoverContextValue } from '@fluentui/react-popover';
import { TeachingPopoverCarouselNavProps } from '../TeachingPopoverCarouselNav/index';

export type TeachingPopoverCarouselFooterSlots = {
  /**
   * The element wrapping carousel pages and navigation.
   */
  root: NonNullable<Slot<'div'>>;

  /**
   * The previous button slot.
   */
  previous?: Slot<typeof Button>;

  /**
   * The next button slot.
   */
  next: NonNullable<Slot<typeof Button>>;

  /**
   * The page nav slot when using icon navigation.
   */
  nav?: Slot<TeachingPopoverCarouselNavProps>;

  /**
   * The page count text slot for paginationType text
   */
  pageCount?: Slot<'div'>;
};

export type TeachingPopoverCarouselFooterLayout = 'offset' | 'centered';

// For localization or customization, users may want to modify this for their own purposes
export type TeachingPopoverPageCountChildRenderFunction = (currentPage: number, totalPages: number) => React.ReactNode;

/**
 * TeachingPopoverCarouselFooter Props
 */
export type TeachingPopoverCarouselFooterProps = ComponentProps<TeachingPopoverCarouselFooterSlots> & {
  /**
   * Controls whether buttons will be centered (balanced) or right aligned
   * Defaults to 'centered'.
   */
  layout?: TeachingPopoverCarouselFooterLayout;

  /**
   * Dictates whether pagination uses text or icons
   * Defaults to icons
   */
  paginationType?: 'text' | 'icon';

  /**
   * The text to be displayed on the initial step of carousel
   */
  initialStepText: string;

  /**
   * The text to be displayed on the final step of carousel
   */
  finalStepText: string;

  /**
   * Localize the page count text via function to fully override
   */
  renderPageCountText?: TeachingPopoverPageCountChildRenderFunction;
};

/**
 * TeachingPopoverCarouselFooter State and Context Hooks
 */
export type TeachingPopoverCarouselFooterState = ComponentState<Required<TeachingPopoverCarouselFooterSlots>> &
  Partial<Pick<PopoverContextValue, 'appearance'>> &
  Pick<TeachingPopoverCarouselFooterProps, 'layout'>;
