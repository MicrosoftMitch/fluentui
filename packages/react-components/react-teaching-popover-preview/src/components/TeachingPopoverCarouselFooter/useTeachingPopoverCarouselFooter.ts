import { usePopoverContext_unstable } from '@fluentui/react-popover';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import * as React from 'react';

import type {
  TeachingPopoverCarouselFooterProps,
  TeachingPopoverCarouselFooterState,
} from './TeachingPopoverCarouselFooter.types';
import { TeachingPopoverCarouselFooterButton } from '../TeachingPopoverCarouselFooterButton/TeachingPopoverCarouselFooterButton';

export const useTeachingPopoverCarouselFooter_unstable = (
  props: TeachingPopoverCarouselFooterProps,
  ref: React.Ref<HTMLDivElement>,
): TeachingPopoverCarouselFooterState => {
  const { layout = 'centered' } = props;

  const appearance = usePopoverContext_unstable(context => context.appearance);

  const previous = slot.optional(props.previous, {
    defaultProps: {
      navType: 'prev',
    },
    renderByDefault: true,
    elementType: TeachingPopoverCarouselFooterButton,
  });

  const next = slot.always(props.next, {
    defaultProps: {
      navType: 'next',
    },
    elementType: TeachingPopoverCarouselFooterButton,
  });

  return {
    appearance,
    layout,
    components: {
      root: 'div',
      next: TeachingPopoverCarouselFooterButton,
      previous: TeachingPopoverCarouselFooterButton,
    },
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
    previous,
    next,
  };
};
