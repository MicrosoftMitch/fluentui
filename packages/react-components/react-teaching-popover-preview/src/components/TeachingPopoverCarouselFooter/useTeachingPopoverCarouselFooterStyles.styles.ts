import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type {
  TeachingPopoverCarouselFooterSlots,
  TeachingPopoverCarouselFooterState,
} from './TeachingPopoverCarouselFooter.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens } from '@fluentui/react-theme';

export const teachingPopoverCarouselFooterClassNames: SlotClassNames<TeachingPopoverCarouselFooterSlots> = {
  root: 'fui-TeachingPopoverCarouselFooter',
  previous: 'fui-TeachingPopoverCarouselFooter__previous',
  next: 'fui-TeachingPopoverCarouselFooter__next',
  nav: 'fui-TeachingPopoverCarouselFooter__nav',
  pageCount: 'fui-TeachingPopoverCarouselFooter__pageCount',
};

// Todo: Page change animation & styles
const useStyles = makeStyles({
  root: {
    display: 'flex',
    flexDirection: 'row',
  },
  rootCentered: {
    justifyContent: 'space-between',
    ...shorthands.gap('8px'),
  },
  rootRightAligned: {
    ...shorthands.gap('8px'),
    '& :first-child': {
      marginInlineEnd: 'auto',
    },
  },
  previous: {
    minWidth: '96px',
  },
  next: {
    minWidth: '96px',
  },
  brandNext: {
    color: tokens.colorBrandForeground1,
    backgroundColor: tokens.colorNeutralForegroundOnBrand,
    ...shorthands.borderColor(tokens.colorTransparentBackground),
    ':hover': {
      color: tokens.colorCompoundBrandForeground1Hover,
      backgroundColor: tokens.colorNeutralForegroundOnBrand,
    },
    ':hover:active': {
      color: tokens.colorCompoundBrandForeground1Pressed,
      backgroundColor: tokens.colorNeutralForegroundOnBrand,
    },
  },
  brandPrevious: {
    // In brand, this is always 'NeutralForegroundOnBrand'
    color: tokens.colorNeutralForegroundOnBrand,
    backgroundColor: tokens.colorBrandBackground,
    ...shorthands.borderColor(tokens.colorNeutralForegroundOnBrand),
    ':hover': {
      color: tokens.colorNeutralForegroundOnBrand,
      ...shorthands.borderColor(tokens.colorNeutralForegroundOnBrand),
      backgroundColor: tokens.colorBrandBackgroundHover,
    },
    ':hover:active': {
      color: tokens.colorNeutralForegroundOnBrand,
      ...shorthands.borderColor(tokens.colorNeutralForegroundOnBrand),
      backgroundColor: tokens.colorBrandBackgroundPressed,
    },
  },
  nav: {},
  pageCount: { display: 'flex', color: tokens.colorNeutralForeground3, justifyContent: 'center', alignItems: 'center' },
});

/** Applies style classnames to slots */
export const useTeachingPopoverCarouselFooterStyles_unstable = (state: TeachingPopoverCarouselFooterState) => {
  const styles = useStyles();
  const { appearance, layout } = state;

  state.root.className = mergeClasses(
    teachingPopoverCarouselFooterClassNames.root,
    styles.root,
    layout === 'centered' ? styles.rootCentered : styles.rootRightAligned,
    state.root.className,
  );
  if (state.previous) {
    state.previous.className = mergeClasses(
      teachingPopoverCarouselFooterClassNames.previous,
      styles.previous,
      appearance === 'brand' && styles.brandPrevious,
      state.previous.className,
    );
  }
  state.next.className = mergeClasses(
    teachingPopoverCarouselFooterClassNames.next,
    styles.next,
    appearance === 'brand' && styles.brandNext,
    state.next.className,
  );
  if (state.nav) {
    state.nav.className = mergeClasses(teachingPopoverCarouselFooterClassNames.nav, styles.nav, state.nav.className);
  }
  if (state.pageCount) {
    state.pageCount.className = mergeClasses(
      teachingPopoverCarouselFooterClassNames.pageCount,
      styles.pageCount,
      state.pageCount.className,
    );
  }

  return state;
};
