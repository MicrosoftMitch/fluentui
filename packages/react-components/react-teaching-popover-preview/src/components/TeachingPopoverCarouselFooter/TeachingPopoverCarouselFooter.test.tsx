import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { TeachingPopoverCarouselFooter } from './TeachingPopoverCarouselFooter';

import { resetIdsForTests } from '@fluentui/react-utilities';

// jest.mock('../TeachingPopoverCarousel/Carousel/CarouselContext');

describe('TeachingPopoverCarouselFooter', () => {
  isConformant({
    Component: TeachingPopoverCarouselFooter,
    displayName: 'TeachingPopoverCarouselFooter',
    requiredProps: {
      next: 'Next',
      previous: 'Previous',
      initialStepText: 'Close',
      finalStepText: 'Finish',
    },
    disabledTests: ['component-has-static-classnames-object'],
  });

  beforeEach(() => {
    resetIdsForTests();
  });

  it('renders a default state', () => {
    const result = render(
      <TeachingPopoverCarouselFooter
        next="Next"
        previous="Previous"
        initialStepText="Close"
        finalStepText="Finish"
        pageCount="of"
      >
        Default TeachingPopoverCarouselFooter
      </TeachingPopoverCarouselFooter>,
    );
    expect(result.container).toMatchSnapshot();
  });
});
