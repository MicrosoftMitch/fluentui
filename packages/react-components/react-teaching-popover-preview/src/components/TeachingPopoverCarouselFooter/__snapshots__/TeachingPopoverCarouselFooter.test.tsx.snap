// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingPopoverCarouselFooter renders a default state 1`] = `
<div>
  <div
    class="fui-TeachingPopoverCarouselFooter"
  >
    <button
      class="fui-Button fui-TeachingPopoverCarouselFooter__previous"
      type="button"
    >
      Close
    </button>
    <div
      class="fui-TeachingPopoverCarouselNav fui-TeachingPopoverCarouselFooter__nav"
      data-tabster="{\\"mover\\":{\\"cyclic\\":false,\\"direction\\":2,\\"memorizeCurrent\\":false,\\"hasDefault\\":true}}"
      role="tablist"
      tabindex="0"
    />
    <button
      class="fui-Button fui-TeachingPopoverCarouselFooter__next"
      type="button"
    >
      Next
    </button>
  </div>
</div>
`;
